lib/rules/id-match.js

```diff
@@ -76,14 +83,23 @@ module.exports = {
         const ALLOWED_PARENT_TYPES = new Set(["CallExpression", "NewExpression"]);
         const DECLARATION_TYPES = new Set(["FunctionDeclaration", "VariableDeclarator"]);
         const IMPORT_TYPES = new Set(["ImportSpecifier", "ImportNamespaceSpecifier", "ImportDefaultSpecifier"]);
+        const BUILT_IN_REFERENCES = new Set(Object.keys(globals.builtin));
 
         /**
-         * Checks if a string matches the provided pattern
+         * Checks if a string matches the provided pattern and is not a global built-in reference
          * @param {string} name The string to check.
          * @returns {boolean} if the string is a match
          * @private
          */
         function isInvalid(name) {
+            const scope = context.getScope();
+            const variable = astUtils.getVariableByName(scope, name);
+
+            // Do not report global built-in references
+            if (variable && variable.scope.type === "global" && BUILT_IN_REFERENCES.has(name)) {
+                return false;
+            }
+
```

Comment:
This can mistakenly allow properties, which are not references.

```js
/* eslint id-match: ["error", "^[^A]", { properties: true }] */

Array.from(foo); // no error, correct

foo.Array = 5; // no error, but it should be
```

We should check if the Identifier node is really a reference.

Also, this rule should ignore all globals except for those that are declared in the same file, as it aims to enforce naming conventions, If a global variable isn't declared in the same file, then we can assume that user has no control over its naming - it can be a native (ES builtin) global, but also an environment variable (e.g., a browser global), framework global, a variable created by a library, etc, so we shouldn't use `globals.builtin`.

The logic can be same as in `id-denylist`:

https://github.com/eslint/eslint/blob/eafaf52464c01072fcf7be2bd2d91cd0b3ae67d0/lib/rules/id-denylist.js#L137-L148

Reply:
I have updated the code to use `isReferenceToGlobalVariable` 👍🏻 , should we move this function to `ast-utils` to avoid duplicate code?

Reply:
> should we move this function to `ast-utils` to avoid duplicate code

I think it's fine to have this function in rules, even though it's duplicate, because it doesn't look generic enough for a shared place. In particular, returning `false` for globals that are declared in the same file is logic that is specific for these rules, so we would probably want to control that with a parameter, and we'd certainly need another parameter for scope, so that already seems too complicated for a few lines of code.